### DISCUSSION TITLE
Compile fast binaries using GraalVM

### DISCUSSION THREAD

**Discussion by bartkl** [beetbox/beets#4560] [Ideas]
Beets is still fantastic, but I would love for it to improve in terms of speed. Would it be worth looking into compiling using GraalVM?

> On average, Python in GraalVM is 8.92x faster than CPython and 8.34x faster than Jython
(Source: https://www.graalvm.org/python/)

I'm not savvy enough with compiler stuff to judge this properly. Would this be a nice thing to look into?

**Comment by wisp3rwind** [comment]
I doubt that this would lead to any substantial improvements: Compute-heavy tasks in beets (converting, replaygain, etc) should all be performed by external programs/libraries which should usually be written in compiled languages. In many other cases, beets is in fact IO-bound due to waiting for web services to fetch data. Another long-standing bottleneck is inefficient database access patterns, but solving that  requires architectural changes (in fact, see https://github.com/beetbox/beets/pull/4362, which I should really come back to review), and another interpreter is probably not going to cut it.

That being said, I don't know GraalVM at all, so by all means, feel free to try this!

(Another reason not to do this would be that GraalVM is certainly not going to be as ubiquitous as CPython.)

**Reply by sampsyo** [reply]
I agree on both points here—if anybody's interested in giving it a shot just for fun, I would be curious to hear the results! But I do think the performance bottlenecks are probably elsewhere. If we want to try to improve performance, I think doing some profiling work might be a more bang-for-buck way to spend time. (It would also be a nice excuse to try the new [Scalene](https://github.com/plasma-umass/scalene) profiler…)

**Reply by bartkl** [reply]
Thanks for the feedback.

And I think trying out that profiler is indeed the way to go.
